src/inventory/transaction/EnchantTransaction.php

```diff
@@ -0,0 +1,98 @@
+<?php
+
+/*
+ *
+ *  ____            _        _   __  __ _                  __  __ ____
+ * |  _ \ ___   ___| | _____| |_|  \/  (_)_ __   ___      |  \/  |  _ \
+ * | |_) / _ \ / __| |/ / _ \ __| |\/| | | '_ \ / _ \_____| |\/| | |_) |
+ * |  __/ (_) | (__|   <  __/ |_| |  | | | | | |  __/_____| |  | |  __/
+ * |_|   \___/ \___|_|\_\___|\__|_|  |_|_|_| |_|\___|     |_|  |_|_|
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * @author PocketMine Team
+ * @link http://www.pocketmine.net/
+ *
+ *
+ */
+
+declare(strict_types=1);
+
+namespace pocketmine\inventory\transaction;
+
+use pocketmine\block\inventory\EnchantInventory;
+use pocketmine\inventory\transaction\action\InventoryAction;
+use pocketmine\item\Item;
+use pocketmine\item\ItemTypeIds;
+use pocketmine\player\Player;
+use function count;
+
+class EnchantTransaction extends InventoryTransaction{
+
+	private int $optionId;
+	private ?Item $inputItem = null;
+	private int $lapisCost = 0;
+	private int $enchantmentLevel = 0;
+
+	public function __construct(Player $source, int $optionId){
+		$this->optionId = $optionId;
+		parent::__construct($source);
+	}
+
+	public function validate() : void{
+		if(count($this->actions) < 1){
+			throw new TransactionValidationException("Transaction must have at least one action to be executable");
+		}
+
+		$enchantWindow = $this->source->getCurrentWindow();
+		if(!$enchantWindow instanceof EnchantInventory){
+			throw new TransactionValidationException("Current window was expected to be of type EnchantInventory");
+		}
+
+		$enchantmentLevel = $enchantWindow->getOptionEnchantmentLevel($this->optionId);
+		if($enchantmentLevel === null){
+			throw new TransactionValidationException("Incorrect option id $this->optionId");
+		}
+		$this->enchantmentLevel = $enchantmentLevel;
```

Comment:
Probably better to take the enchantment option itself here, so that we can give it to `EnchantItemEvent` for plugins to see